### PR TITLE
[WIP] Refactor/keybindings

### PR DESCRIPTION
**Pull request by geksilla**
This PR not ready to merge.

The goals of refactoring are:
- add more control to keybindings associated with different layers, modes and packages
- allow to bind keystrokes for `ctrl-`, `alt-` etc.
- split keybindings to `keymaps` storage like `mode-keymap`, `proton-keymap`, `atom-keymap` etc.
- manage keybindings according to package/layer state.
- provide core functions to define keybindings for package, layer, proton.

For now implemented:
- different keymap storage. `proton-keymap` holds all keybindings to be executed when proton-mode active. `mode-keymap` holds all mode keybindings.
- convert existing hash-map formated keybindings to new format to not break existing layer keybindings and users' in `.proton`.
- colorized proton bottom panel and format titles by types like in spacemacs (works only on `SPC` for now). 
bottom_panel
- `proton:chain` command keymaps injected dynamically to capture all possible input keystrokes.

### CODE REVIEW

**Comment by dvcrn** [comment]
Really cool! Also love the new style you went with :+1: 
If I can help somehow, let me know!

**Comment by geksilla** [comment]
Thanks @dvcrn!

**Comment by geksilla** [comment]
@dvcrn, now bottom panel colorized and chaining works . You can try it. Mode keybindings panel doesn't work for now.

**Comment by geksilla** [comment]
Mode keybindings now works as well by `SPC m` and `,` alias.

**Comment by spencerlyon2** [comment]
This looks like an awesome set of changes. I'm hoping I find some time to
play around with this tonight or tomorrow

// Spencer

**Comment by geksilla** [comment]
Initial work is done. New keybindings manager should work fine with existing hash-map format, so there are no breaking changes regarding layers and `~/.proton` files format. I will continue my work for additional function regarding managing package and mode keybindings and find the way how to link between them. I'm experimenting with DSL functions for layers. Basically this PR can be merged I will split my future work in few PRs. Ping @dvcrn, @spencerlyon2 to review and test. Thanks.

**Comment by dvcrn** [comment]
I'm looking through this right now. Why did you decide to go with string combinations instead of a prefix tree? 

> add more control to keybindings associated with different layers, modes and packages

what kind of control is missing?

**Comment by geksilla** [comment]
> what kind of control is missing?
I want to add hooks which will help to define keybindings linked to package. So we could remove/add keybindings when package state changed.
> Why did you decide to go with string combinations instead of a prefix tree?
Good question. Actualy I missed that prefix tree would be more efficient and easier to filter through. I will try to change this today.

**Comment by geksilla** [comment]
@dvcrn, updated keymaps to prefix tree. I can squash commits if needed. Thanks.

**Comment by dvcrn** [comment]
That was fast. I was just wondering and didn't mean to say that you have to completely change it :innocent: 

This PR looks good from my side. Waiting for @spencerlyon2

**Comment by geksilla** [comment]
One more commit, fixed minor issue with punctuation chars capturing.

**Comment by spencerlyon2** [comment]
Sorry to hold things up. I'll be able to look at this tonight (5-7 hours). If you're eager to merge and both feel good about it, I'm sure it's a quality change

File: plugin/keymaps/proton.cson [hunk 1]
```diff
@@ -16,76 +16,76 @@
...
-  '5': 'proton:chain'
-  '6': 'proton:chain'
-  '7': 'proton:chain'
-  '8': 'proton:chain'
-  '9': 'proton:chain'
-  '0': 'proton:chain'
-  'escape': 'proton:chain'
-  'backspace': 'proton:chain'
-  'tab': 'proton:chain'
+# 'atom-text-editor.proton-mode':
```
Comment (spencerlyon2) [inline plugin/keymaps/proton.cson]: I'm in favor of deleting the commented out lines

File: src/cljs/proton/lib/helpers.cljs [hunk 2]
```diff
@@ -54,19 +60,19 @@
...
 
+(defn keybinding-row-html [keybinding]
+  (let [options (nth keybinding 1)
+        {:keys [category action title]} options
+        is-category? ((comp not nil?) category)
+        class-name (if is-category? "proton-key-category" "proton-key-action")
+        value (if is-category? (str "+" category) (or title action))
+        keystroke (name (key keybinding))]
+      (str "<li class='flex-item'><span class='proton-key'>[" keystroke "]</span> ➜ <span class='" class-name "'>" value "</span></li>")))
+
```
Comment (spencerlyon2) [inline src/cljs/proton/lib/helpers.cljs]: Awesome! Good work

File: src/cljs/proton/lib/proton.cljs [hunk 3]
```diff
@@ -56,3 +57,18 @@
...
+(defn init-proton-mode-keymaps! []
+  (let [selector "atom-text-editor.proton-mode"
+        command "proton:chain"
+        make-char-fn (comp (partial map char) range)
+        letters (make-char-fn 97 123)
+        numbers (map str (range 0 9))
+        punctuation (flatten (map #(apply make-char-fn %) [[33 48] [58 65] [91 97] [123 127]]))
+        special-chars ["escape" "tab" "backspace" "delete"]
+        all-chars (into [] (concat [] letters (map string/upper-case letters) numbers special-chars punctuation))
+        combo ["ctrl-" "ctrl-alt-" "ctrl-shift-" "ctrl-shift-alt-" "alt-" "alt-shift-"]
```
Comment (spencerlyon2) [inline src/cljs/proton/lib/proton.cljs]: What about `cmd` for OSX users? Do we need them here too?

**Comment by spencerlyon2** [comment]
So I've taken a look and I think these are all great changes. I did make that line comment about including `cmd` in the combos section of the keymap init. After that I think we can merge

**Comment by geksilla** [comment]
thanks @spencerlyon2. Nice catch for `cmd`. Will add it to combos.

**Comment by geksilla** [comment]
Added `cmd` to combos.

**Comment by dvcrn** [comment]
Merged :+1:

**Pull request merged**